src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -413,19 +413,35 @@ protected function instantiateObject(array &$data, string $class, array &$contex
                         throw new MissingConstructorArgumentsException(sprintf('Cannot create an instance of "%s" from serialized data because its constructor requires parameter "%s" to be present.', $class, $constructorParameter->name), 0, null, [$constructorParameter->name]);
                     }
 
+                    $expectedTypes = ['unknown'];
+
+                    if ($constructorParameter->hasType()) {
+                        if (\PHP_VERSION_ID >= 80000) {
+                            if ($constructorParameter->getType() instanceof \ReflectionUnionType) {
+                                $expectedTypes = array_map(function ($type) { return $type->getName(); }, $constructorParameter->getType()->getTypes());
+                            } else {
+                                $expectedTypes = [$constructorParameter->getType()->getName()];
+                            }
```

Comment:
In PHP8.1, there is a new "intersection" type. And the `ReflectionIntersectionType` do not have any `getName` method.
Therefore IMHO, the easiest way to go is to rely on the fact that the `ReflectionPropertyType` is stringable.

So we can do something like that:
```php
if ($constructorParameter->getType() instanceof \ReflectionUnionType) {
    $expectedTypes = array_map(function (\ReflectionType $t): string { return (string) $t; }, $constructorParameter->getType()->getTypes());
} else {
    $expectedTypes = [(string) $constructorParameter->getType()];
}
```
It'll handle every kind of `ReflectionPropertyType` and only split the expected type array whenever it's an union type.